### PR TITLE
feat(datasets): make strict_dataset_uri_validation default to True

### PR DESCRIPTION
**Pull request by Lee-W**
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/datasets/__init__.py [hunk 1]
```diff
@@ -108,15 +108,9 @@ def _sanitize_uri(uri: str) -> str:
     if (normalizer := _get_uri_normalizer(normalized_scheme)) is not None:
         try:
             parsed = normalizer(parsed)
-        except ValueError as exception:
+        except ValueError:
             if conf.getboolean("core", "strict_dataset_uri_validation", fallback=False):
```
Comment (hussein-awala) [inline airflow/datasets/__init__.py]: Nit:
```suggestion
            if conf.getboolean("core", "strict_dataset_uri_validation", fallback=True):
```

**Review by hussein-awala — APPROVED** [review]

File: airflow/datasets/__init__.py (hunk 1 above)
Comment (Lee-W) [reply airflow/datasets/__init__.py]: Thanks!

File: newsfragments/41814.significant.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Change the default value of ``strict_dataset_uri_validation`` to True.
```
Comment (uranusjr) [inline newsfragments/41814.significant.rst:1]: We should mention somewhere that non-URI-based datasets should use `name` instead, but that hasn’t been implemented so I’m not sure if it’s a good idea to mention it here. But if we don’t, we might fotget and not mentioning this at all in the end.

Comment (Lee-W) [reply newsfragments/41814.significant.rst:1]: I feel like it does not really make sense to mention here 🤔 this one is not actually blocking anything. or maybe we could wait for that to be implemented and then merge this one (with `using name` mentioned)

**Comment by Lee-W** [comment]
https://cwiki-test.apache.org/confluence/display/AIRFLOW/AIP-60+Standard+URI+representation+for+Airflow+Datasets

File: newsfragments/41814.significant.rst (hunk 2 above)
Comment (uranusjr) [reply newsfragments/41814.significant.rst:1]: Discussed with Wei earlier. This error only matters when you use certain URI schemes that have specific semantics (e.g. Postgres, S3).  fter AIP-72, you still need to provide a URI when the asset is named _when this error message matters_ (the entire point if for you to provide a meaningful URI. So the better way to help users migrate would be to

1. Make this an error (already done in this PR)
2. Add documentation and add a link in the error message to point users to the right directions.
    * If you can, provide a good URI for the resource. The `name` argument can optionally be used to provide a more human-readable identifier to the asset.
    * If you can’t provide a meaningful URI, use another scheme altogether that don’t have the semantic restrictions. Airflow will never require a semantic for user-defined URI schemes (with a prefix `x-`), so that can be a good alternative.
    * If you have a URI, but it can only be obtained later (at task execution), consider using `AssetAlias` instead, and provide the URI later.

Comment (Lee-W) [reply newsfragments/41814.significant.rst:1]: I just updated it along with some documentation fix

File: tests/io/test_path.py [hunk 3]
```diff
@@ -406,7 +406,7 @@ def test_asset(self):
         attach("s3", fs=FakeRemoteFileSystem())
 
         p = "s3"
-        f = "/tmp/foo"
+        f = "tmp/foo"
```
Comment (uranusjr) [inline tests/io/test_path.py]: Why the change here?

Comment (Lee-W) [reply tests/io/test_path.py]: it should be `s3://../..` instead of `s3:///../..`

Comment (uranusjr) [reply tests/io/test_path.py]: Ah I see. We should do this instead to make it more obvious

```suggestion
        f = "bucket/object"
```

**Review by uranusjr — APPROVED** [review]

File: tests/io/test_path.py (hunk 3 above)
Comment (Lee-W) [reply tests/io/test_path.py]: sounds like a great idea!
